src/libopensc/card-sc-hsm.c

```diff
@@ -1424,10 +1432,190 @@ static int sc_hsm_unwrap_key(sc_card_t *card, sc_cardctl_sc_hsm_wrapped_key_t *p
 	r = sc_transmit_apdu(card, &apdu);
 	LOG_TEST_RET(ctx, r, "APDU transmit failed");
 
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(ctx, r, "Check SW error");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+
+static int get_CAR(u8 *carstr, sc_context_t *ctx, const u8 *buf, size_t buflen)
+{
+	const u8 *cb = NULL, *car = NULL;
+	size_t taglen;
+
+	LOG_FUNC_CALLED(ctx);
+
+	/* find embedded Certificate Body (should be right at the start) */
+	if (!(cb = sc_asn1_find_tag(ctx, buf, buflen, 0x7F4E, &taglen))) {
+		LOG_FUNC_RETURN(ctx, SC_ERROR_UNKNOWN);
+	}
+	buf = cb;
+	buflen = taglen;
+
+	/* find embedded Certification Authority Reference (CAR) */
+	if (!(car = sc_asn1_find_tag(ctx, buf, buflen, 0x42, &taglen))) {
+		LOG_FUNC_RETURN(ctx, SC_ERROR_UNKNOWN);
+	}
+
+	/* return CAR */
+	memcpy(carstr, car, taglen);
```

Comment:
possible out of bounds write. you need to check taglen against the length of your target buffer.

Reply:
Fixed, thanks.